runtime/syntax/swig.vim

```diff
@@ -0,0 +1,99 @@
+" Vim syntax file
+" Language:	SWIG
+" Maintainer:	Julien Marrec <julien.marrec 'at' gmail com>
+" Last Change:	2023 November 23
+
+if exists("b:current_syntax")
+  finish
+endif
+
+" Read the C++ syntax to start with
+runtime! syntax/cpp.vim
+unlet b:current_syntax
+
+" SWIG extentions
+syn keyword swigInclude %include %import %importfile %includefile %module
+
+syn keyword swigMostCommonDirective %alias %apply %beginfile %clear %constant %define %echo %enddef %endoffile
+syn keyword swigMostCommonDirective %extend %feature %director %fragment %ignore  %inline
+syn keyword swigMostCommonDirective %keyword %name %namewarn %native %newobject %parms %pragma
+syn keyword swigMostCommonDirective %rename %template %typedef %typemap %types %varargs
+
+" SWIG: Language specific macros
+syn keyword swigOtherLanguageSpecific %luacode %go_import
+
+syn keyword swigCSharp %csattributes %csconst %csconstvalue %csmethodmodifiers %csnothrowexception
+syn keyword swigCSharp %dconstvalue %dmanifestconst %dmethodmodifiers
+
+syn keyword swigJava %javaconstvalue %javaexception %javamethodmodifiers %javaconst %nojavaexception
+
+syn keyword swigGuile %multiple_values %values_as_list %values_as_vector
+
+syn keyword swigPHP %rinit %rshutdown %minit %mshutdown
+
+syn keyword swigPython %pybinoperator %pybuffer_binary %pybuffer_mutable_binary %pybuffer_mutable_string %pybuffer_string
+syn keyword swigPython %pythonappend %pythonbegin %pythoncode %pythondynamic %pythonnondynamic %pythonprepend
+
+syn keyword swigRuby %markfunc %trackobjects %bang
+syn keyword swigScilab %scilabconst
+
+" SWIG: Insertion
+syn keyword swigInsertSection %insert %begin %runtime %header %wrapper %init
+
+" SWIG: Other directives
+syn keyword swigCstring %cstring_bounded_mutable %cstring_bounded_output %cstring_chunk_output %cstring_input_binary %cstring_mutable
+syn keyword swigCstring %cstring_output_allocate %cstring_output_allocate_size %cstring_output_maxsize %cstring_output_withsize
+syn keyword swigCWstring %cwstring_bounded_mutable %cwstring_bounded_output %cwstring_chunk_output %cwstring_input_binary %cwstring_mutable
+syn keyword swigCWstring %cwstring_output_allocate %cwstring_output_allocate_size %cwstring_output_maxsize %cwstring_output_withsize
+syn keyword swigCMalloc %malloc %calloc %realloc %free %sizeof %allocators
+
+syn keyword swigExceptionHandling %catches %raise %allowexception %exceptionclass %warn %warnfilter %exception
+syn keyword swigContract %contract %aggregate_check
+
+syn keyword swigDirective %addmethods %array_class %array_functions %attribute %attribute2 %attribute2ref
```

Comment:
> Sorry for not looking at this sooner, but I notice there are a few long obsolete and deprecated macros included which will be gone completely in the next SWIG release (4.2.0): 

vim should be able to work with old SWIG files as well, shouldn't it?

Omissions are real issue.

Reply:
> vim should be able to work with old SWIG files as well, shouldn't it?

I think in these three cases it is more unhelpful than helpful to treat them as valid syntax.

`%addmethods` was renamed to `%extend` (because it can add member variables too) in SWIG 1.3.12 (June 2, 2002) and has emitted a deprecation warning since then.

`%name` has emitted a deprecation warning since SWIG 1.3.24 (December 14, 2004).  It has to appear right before the thing you want to rename so doesn't work if you want to get SWIG to parse a library's API headers, which is the typical way to use SWIG.  The replacement is `%rename` which can rename things declared in an API header.

You might stick with a deprecated directive while you want to stay compatible with releases before the replacement was added, but I can't see that anyone would have been ignoring a deprecation warning for 19+ years, and I'd argue it's more helpful if they don't highlight as valid directives to give a visual signal that this needs updating if you do have an ancient interface file.

`%attribute_ref` is perhaps less clear cut - it was deprecated in SWIG 1.3.32 (November 15, 2007) but without a deprecation warning.

The replacement is `%attributeref` (with no underscore, which is more consistent with other macro names; also the parameter order now matches the other attribute macros).  I'd still argue for not highlighting it as valid because the old deprecated name is an easy typo to make for the new one.

Reply:
@ojwb I realized some of these were deprecated, but I haven't tried to look at all of these.

I couldn't find a definitive, easy to consume, source for the accepted keywords. So basically what I did was using python regexes to parse through the entire swig/swig repo and trying to derive the keywords, a little bit of manual cleaning after that, and trying to quickly see if some of them were language specific to classify them a bit more. 

If you want to adjust the vim syntax file with your intricate knowledge of swig internals that'd be awesome, and I'm happy to review.

Reply:
I agree that if something has been deprecated for so many years, it's probably best to not highlight them.

Reply:
Sadly SWIG's docs are patchy in places.  We now require docs and tests for new features, but in the past that wasn't always the case and it's a bit of a daunting task to try to fill in missing docs.  I've been involved in SWIG dev since ~2005 and I've little idea about some of these features and sometimes all there is to go on is a brief mention in CHANGES, or maybe nothing at all.

The next release will pretty much remove everything marked as deprecated more than about 10 years ago which at least reduces the number of such things.

I can't really think of a good way to generate such a list.  It doesn't help that some of these are built in features handle by SWIG's parser whereas others are defined as macros in the library (the user mostly doesn't need to know or care about the difference but it makes generating a list of them more challenging).

I'll certainly feedback improvements I spot.